registry/config.go

```diff
@@ -67,7 +67,7 @@ var lookupIP = net.LookupIP
 // InstallCliFlags adds command-line options to the top-level flag parser for
 // the current process.
 func (options *ServiceOptions) InstallCliFlags(flags *pflag.FlagSet) {
-	mirrors := opts.NewNamedListOptsRef("registry-mirrors", &options.Mirrors, ValidateMirror)
```

Comment:
Actually, perhaps we should update those error messages a bit, as they are a bit inconsistent, and don't show _which_ mirror/url was invalid;

```go
// ValidateMirror validates an HTTP(S) registry mirror
func ValidateMirror(val string) (string, error) {
	uri, err := url.Parse(val)
	if err != nil {
		return "", fmt.Errorf("invalid mirror: %q is not a valid URI", val)
	}
	if uri.Scheme != "http" && uri.Scheme != "https" {
		return "", fmt.Errorf("invalid mirror: unsupported scheme %q in %q", uri.Scheme, uri)
	}
	if (uri.Path != "" && uri.Path != "/") || uri.RawQuery != "" || uri.Fragment != "" {
		return "", fmt.Errorf("invalid mirror: path, query, or fragment at end of the URI %q", uri)
	}
	return strings.TrimSuffix(val, "/") + "/", nil
}
```

```patch
diff --git a/registry/config.go b/registry/config.go
index 79819de..88b62af 100644
--- a/registry/config.go
+++ b/registry/config.go
@@ -247,18 +247,15 @@ func isSecureIndex(config *serviceConfig, indexName string) bool {
 func ValidateMirror(val string) (string, error) {
        uri, err := url.Parse(val)
        if err != nil {
-               return "", fmt.Errorf("%s is not a valid URI", val)
+               return "", fmt.Errorf("invalid mirror: %q is not a valid URI", val)
        }
-
        if uri.Scheme != "http" && uri.Scheme != "https" {
-               return "", fmt.Errorf("Unsupported scheme %s", uri.Scheme)
+               return "", fmt.Errorf("invalid mirror: unsupported scheme %q in %q", uri.Scheme, uri)
        }
-
-       if uri.Path != "" || uri.RawQuery != "" || uri.Fragment != "" {
-               return "", fmt.Errorf("Unsupported path/query/fragment at end of the URI")
+       if (uri.Path != "" && uri.Path != "/") || uri.RawQuery != "" || uri.Fragment != "" {
+               return "", fmt.Errorf("invalid mirror: path, query, or fragment at end of the URI %q", uri)
        }
-
-       return fmt.Sprintf("%s://%s/", uri.Scheme, uri.Host), nil
+       return strings.TrimSuffix(val, "/") + "/", nil
 }
```